[JSC] Move Promise.all implementation to C++

Constellation opened the pull request:
#### 824017ee2ca01569f75426e73345bc9f7cb95cf8
<pre>
[JSC] Move Promise.all implementation to C++
<a href="https://bugs.webkit.org/show_bug.cgi?id=303603">https://bugs.webkit.org/show_bug.cgi?id=303603</a>
<a href="https://rdar.apple.com/165891602">rdar://165891602</a>

Reviewed by Sosuke Suzuki.

JS version of that has serious performance issue for the startup case.
Once it gets FTL, it gets much better. But Baseline / DFG are really
slow compared to C++ native code.

1. We move all of Promise.all code from JS to C++. We reimplement it in
   the same way to what Promise.race is doing. We leverage existing
   JSPromise related helper functions so not so many new code is
   necessary.
2. We introduce `PromiseAllResolveJob` InternalMicrotask which is used
   for fast path Promise.all driving function. We can avoid passing a
   function and we can fully execute these reaction job in C++ in
   runInternalMicrotask.
3. We make JSPromiseAllContext / JSPromiseAllGlobalContext just cells
   instead of objects since they are no longer exposed to builtin JS.
   Also we can remove many of DFG / FTL handlings for these objects
   since they are no longer necessary.

* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/PromiseConstructor.js:
(linkTimeConstant.promiseAllSlow): Deleted.
(linkTimeConstant.promiseNewOnRejected): Deleted.
(linkTimeConstant.promiseAllNewResolveElement): Deleted.
(all): Deleted.
* Source/JavaScriptCore/bytecode/BytecodeIntrinsicRegistry.h:
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.h:
(JSC::BytecodeGenerator::emitIsPromise):
(JSC::BytecodeGenerator::emitIsPromiseAllContext): Deleted.
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::promiseAllContextInternalFieldIndex): Deleted.
(JSC::promiseAllGlobalContextInternalFieldIndex): Deleted.
(JSC::BytecodeIntrinsicNode::emit_intrinsic_getPromiseAllContextInternalField): Deleted.
(JSC::BytecodeIntrinsicNode::emit_intrinsic_getPromiseAllGlobalContextInternalField): Deleted.
(JSC::BytecodeIntrinsicNode::emit_intrinsic_putPromiseAllContextInternalField): Deleted.
(JSC::BytecodeIntrinsicNode::emit_intrinsic_putPromiseAllGlobalContextInternalField): Deleted.
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGObjectAllocationSinkingPhase.cpp:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNewInternalFieldObject):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/ftl/FTLOperations.cpp:
(JSC::FTL::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/JSFunctionWithFields.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
(JSC::JSGlobalObject::visitChildrenImpl):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::wrapForValidIteratorStructure const):
(JSC::JSGlobalObject::promiseAllContextStructure const): Deleted.
(JSC::JSGlobalObject::promiseAllGlobalContextStructure const): Deleted.
* Source/JavaScriptCore/runtime/JSMicrotask.cpp:
(JSC::runInternalMicrotask):
* Source/JavaScriptCore/runtime/JSPromise.cpp:
* Source/JavaScriptCore/runtime/JSPromiseAllContext.cpp:
(JSC::JSPromiseAllContext::create):
(JSC::JSPromiseAllContext::visitChildrenImpl):
(JSC::JSPromiseAllContext::createWithInitialValues): Deleted.
(JSC::JSPromiseAllContext::finishCreation): Deleted.
(JSC::JSC_DEFINE_HOST_FUNCTION): Deleted.
* Source/JavaScriptCore/runtime/JSPromiseAllContext.h:
* Source/JavaScriptCore/runtime/JSPromiseAllGlobalContext.cpp:
(JSC::JSPromiseAllGlobalContext::create):
(JSC::JSPromiseAllGlobalContext::createStructure):
(JSC::JSPromiseAllGlobalContext::visitChildrenImpl):
(JSC::JSPromiseAllGlobalContext::createWithInitialValues): Deleted.
(JSC::JSPromiseAllGlobalContext::finishCreation): Deleted.
(JSC::JSC_DEFINE_HOST_FUNCTION): Deleted.
* Source/JavaScriptCore/runtime/JSPromiseAllGlobalContext.h:
* Source/JavaScriptCore/runtime/JSPromiseConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::promiseAllSlow):
* Source/JavaScriptCore/runtime/JSPromiseConstructor.h:
* Source/JavaScriptCore/runtime/JSType.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/runtime/JSType.h:
* Source/JavaScriptCore/runtime/Microtask.h:
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::VM):
(JSC::VM::promiseAllFulfillFunctionExecutableSlow):
(JSC::VM::promiseAllSlowFulfillFunctionExecutableSlow):
(JSC::VM::visitAggregateImpl):
* Source/JavaScriptCore/runtime/VM.h:
(JSC::VM::promiseAllFulfillFunctionExecutable):
(JSC::VM::promiseAllSlowFulfillFunctionExecutable):

Canonical link: <a href="https://commits.webkit.org/304039@main">https://commits.webkit.org/304039@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/20660029df74c5c39ba795717930c2a3fd9e3bae

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134225 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6739 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45444 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141808 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86282 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/aabcd384-2d13-4dbe-94d0-17e3ef6816cd) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7342 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6601 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102655 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/69913 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137172 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5129 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120312 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83448 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5005 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2586 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/1622 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/126304 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/114201 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38450 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144453 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/132741 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6409 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39031 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/111020 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6492 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5393 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111268 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4813 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116584 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60184 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20750 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6461 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34797 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/165703 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6303 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69927 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43283 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6543 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6413 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

sosukesuzuki commented on Source/JavaScriptCore/runtime/JSPromiseAllGlobalContext.h:
```diff
@@ -65,28 +44,30 @@ class JSPromiseAllGlobalContext final : public JSInternalFieldObjectImpl<JSPromi
-    void finishCreation(VM&, JSValue promise, JSValue values, JSValue remainingElementsCount);
+    WriteBarrier<Unknown> m_promise;
+    WriteBarrier<Unknown> m_values;
+    WriteBarrier<Unknown> m_remainingElementsCount;
```
question: `remainingElementsCount` is always number so can we make this just `uint64_t` ?

Constellation replied:
Currently this data structure is also used in the other place and it needs all of them are JSValues. So we are keeping it as JSValue right now.

sosukesuzuki reviewed (changes requested):
test262 fails? (tested on my local):

```
! NEW FAIL test/built-ins/Promise/all/capability-resolve-throws-no-close.js (strict mode) (Exit code: 3)
Exception: Test262Error:

! NEW FAIL test/built-ins/Promise/all/capability-resolve-throws-no-close.js (default) (Exit code: 3)
Exception: Test262Error:

! NEW FAIL test/built-ins/Promise/all/capability-resolve-throws-reject.js (strict mode) (Exit code: 3)
Exception: Test262Error:

! NEW FAIL test/built-ins/Promise/all/capability-resolve-throws-reject.js (default) (Exit code: 3)
Exception: Test262Error:

! NEW FAIL test/built-ins/Promise/all/resolve-element-function-name.js (strict mode) (Exit code: 3)
Exception: Test262Error: obj should have an own property name

! NEW FAIL test/built-ins/Promise/all/resolve-element-function-name.js (default) (Exit code: 3)
Exception: Test262Error: obj should have an own property name
```

sosukesuzuki reviewed (approved):
r=me, if test262 is green

Constellation committed:
[JSC] Move Promise.all implementation to C++
https://bugs.webkit.org/show_bug.cgi?id=303603
rdar://165891602

Reviewed by Sosuke Suzuki.

JS version of that has serious performance issue for the startup case.
Once it gets FTL, it gets much better. But Baseline / DFG are really
slow compared to C++ native code.

1. We move all of Promise.all code from JS to C++. We reimplement it in
   the same way to what Promise.race is doing. We leverage existing
   JSPromise related helper functions so not so many new code is
   necessary.
2. We introduce `PromiseAllResolveJob` InternalMicrotask which is used
   for fast path Promise.all driving function. We can avoid passing a
   function and we can fully execute these reaction job in C++ in
   runInternalMicrotask.
3. We make JSPromiseAllContext / JSPromiseAllGlobalContext just cells
   instead of objects since they are no longer exposed to builtin JS.
   Also we can remove many of DFG / FTL handlings for these objects
   since they are no longer necessary.

* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/PromiseConstructor.js:
(linkTimeConstant.promiseAllSlow): Deleted.
(linkTimeConstant.promiseNewOnRejected): Deleted.
(linkTimeConstant.promiseAllNewResolveElement): Deleted.
(all): Deleted.
* Source/JavaScriptCore/bytecode/BytecodeIntrinsicRegistry.h:
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.h:
(JSC::BytecodeGenerator::emitIsPromise):
(JSC::BytecodeGenerator::emitIsPromiseAllContext): Deleted.
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::promiseAllContextInternalFieldIndex): Deleted.
(JSC::promiseAllGlobalContextInternalFieldIndex): Deleted.
(JSC::BytecodeIntrinsicNode::emit_intrinsic_getPromiseAllContextInternalField): Deleted.
(JSC::BytecodeIntrinsicNode::emit_intrinsic_getPromiseAllGlobalContextInternalField): Deleted.
(JSC::BytecodeIntrinsicNode::emit_intrinsic_putPromiseAllContextInternalField): Deleted.
(JSC::BytecodeIntrinsicNode::emit_intrinsic_putPromiseAllGlobalContextInternalField): Deleted.
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGObjectAllocationSinkingPhase.cpp:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNewInternalFieldObject):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/ftl/FTLOperations.cpp:
(JSC::FTL::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/JSFunctionWithFields.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
(JSC::JSGlobalObject::visitChildrenImpl):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::wrapForValidIteratorStructure const):
(JSC::JSGlobalObject::promiseAllContextStructure const): Deleted.
(JSC::JSGlobalObject::promiseAllGlobalContextStructure const): Deleted.
* Source/JavaScriptCore/runtime/JSMicrotask.cpp:
(JSC::runInternalMicrotask):
* Source/JavaScriptCore/runtime/JSPromise.cpp:
* Source/JavaScriptCore/runtime/JSPromiseAllContext.cpp:
(JSC::JSPromiseAllContext::create):
(JSC::JSPromiseAllContext::visitChildrenImpl):
(JSC::JSPromiseAllContext::createWithInitialValues): Deleted.
(JSC::JSPromiseAllContext::finishCreation): Deleted.
(JSC::JSC_DEFINE_HOST_FUNCTION): Deleted.
* Source/JavaScriptCore/runtime/JSPromiseAllContext.h:
* Source/JavaScriptCore/runtime/JSPromiseAllGlobalContext.cpp:
(JSC::JSPromiseAllGlobalContext::create):
(JSC::JSPromiseAllGlobalContext::createStructure):
(JSC::JSPromiseAllGlobalContext::visitChildrenImpl):
(JSC::JSPromiseAllGlobalContext::createWithInitialValues): Deleted.
(JSC::JSPromiseAllGlobalContext::finishCreation): Deleted.
(JSC::JSC_DEFINE_HOST_FUNCTION): Deleted.
* Source/JavaScriptCore/runtime/JSPromiseAllGlobalContext.h:
* Source/JavaScriptCore/runtime/JSPromiseConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::promiseAllSlow):
* Source/JavaScriptCore/runtime/JSPromiseConstructor.h:
* Source/JavaScriptCore/runtime/JSType.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/runtime/JSType.h:
* Source/JavaScriptCore/runtime/Microtask.h:
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::VM):
(JSC::VM::promiseAllFulfillFunctionExecutableSlow):
(JSC::VM::promiseAllSlowFulfillFunctionExecutableSlow):
(JSC::VM::visitAggregateImpl):
* Source/JavaScriptCore/runtime/VM.h:
(JSC::VM::promiseAllFulfillFunctionExecutable):
(JSC::VM::promiseAllSlowFulfillFunctionExecutable):

Canonical link: https://commits.webkit.org/304039@main
